Gems/Atom/RHI/Code/Source/RHI/Object.cpp

```diff
@@ -28,5 +28,25 @@ namespace AZ
         {
             return m_name;
         }
+
+        void Object::add_ref() const
+        {
+            AZ_Assert(m_useCount >= 0, "m_useCount is negative");
+            ++m_useCount;
+        }
+
+        void Object::release() const
+        {
+            int useCount = --m_useCount;
+            AZ_Assert(useCount >= 0, "Releasing an already released object");
+
+            if (useCount == 0)
+            {
+                // Get a mutable pointer to shutdown the object before deleting
```

Comment:
suggestion: set m_useCount to a negative value here, force any other accessing thread to assert.

Reply:
It's still expected that after this point, another thread might come and grab a reference to the resource view from the cache, and increment it back to 1. The current code allows that to happen without needing to re-create it. Setting it to -1 here would entail checking for -1 in GetResourceView, and creating a new one in that case instead of using the one that's found, but I figure you might as well re-use what's there.

Though on the other hand, doing would allow this thread to just always delete it, knowing that if another thread tried to find it in the cache, that thread would have to replace it with a new one. So it would simplify the logic of this release a bit.

Reply:
I don't see any extra steps here to validate that another thread hasn't inc'd the use count to keep the object around, so this thread is going to delete it. Better to force other threads to handle that, otherwise there is still a thin gap where 2 threads can disagree what state the object is in. If you try to set the value to -1 and fail because it's no longer zero you will have removed that gap.

Reply:
I see what you're saying. I was conflating ResourceView::release() with this one. The extra step exists for ResourceViews in EraseResourceView. There is no extra step for objects in general. 

This is a good suggestion, as it will prevent other non-resource view systems from running into the same problem (or at least make it more obvious what the problem is), since they should hit the assert in add_ref if they ever try to increment the refcount after the object is on track to be deleted.